File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/SoundFileLoader.java
package com.habitrpg.android.habitica.helpers;

import android.content.Context;
import android.support.annotation.Nullable;

import com.habitrpg.android.habitica.HabiticaApplication;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okio.BufferedSink;
import okio.Okio;
import rx.Observable;
import rx.exceptions.OnErrorThrowable;
import rx.schedulers.Schedulers;

// based on http://stackoverflow.com/questions/29838565/downloading-files-using-okhttp-okio-and-rxjava
public class SoundFileLoader {
    private final Context context;
    private OkHttpClient client;

    public SoundFileLoader(Context context) {
        client = new OkHttpClient();
        this.context = context;
    }

    public Observable<List<SoundFile>> download(List<SoundFile> files) {
        return Observable.from(files)
                .flatMap(audioFile -> {
                    File file = new File(getFullAudioFilePath(audioFile));
                    if (file.exists()) {
                        // Important, or else the MediaPlayer can't access this file
                        file.setReadable(true, false);
                        audioFile.setFile(file);
                        return Observable.just(audioFile);
                    }

                    final Observable<SoundFile> fileObservable = Observable.create(sub -> {
                        Request request = new Request.Builder().url(audioFile.getWebUrl()).build();

                        Response response;
                        try {
                            response = client.newCall(request).execute();
                            if (!response.isSuccessful()) {
                                throw new IOException();
                            }
                        } catch (IOException io) {
                            throw OnErrorThrowable.from(OnErrorThrowable.addValueAsLastCause(io, audioFile));
                        }

                        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
                            try {
                                BufferedSink sink = Okio.buffer(Okio.sink(file));
                                sink.writeAll(response.body().source());
                                sink.flush();
                                sink.close();
                            } catch (IOException io) {
                                throw OnErrorThrowable.from(OnErrorThrowable.addValueAsLastCause(io, audioFile));
                            }

                            file.setReadable(true, false);
                            audioFile.setFile(file);
                            sub.onNext(audioFile);
                            sub.onCompleted();
                        }
                    });
                    return fileObservable.subscribeOn(Schedulers.io());
                }, 5)
                .toList()
                .map(ArrayList::new);
    }

    @Nullable
    private String getExternalCacheDir() {
        File cacheDir = HabiticaApplication.getInstance(context).getExternalCacheDir();
        if (cacheDir != null) {
            return cacheDir.getPath();
        } else {
            return null;
        }
    }

    public String getFullAudioFilePath(SoundFile soundFile) {
        return getExternalCacheDir() + File.separator + soundFile.getFilePath();
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/SoundFileLoader.kt
package com.habitrpg.android.habitica.helpers

import android.annotation.SuppressLint
import android.content.Context
import android.os.Environment
import com.habitrpg.android.habitica.HabiticaBaseApplication
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okio.buffer
import okio.sink
import java.io.File
import java.io.IOException

// based on http://stackoverflow.com/questions/29838565/downloading-files-using-okhttp-okio-and-rxjava
class SoundFileLoader(private val context: Context) {
    private val client: OkHttpClient = OkHttpClient()

    private val externalCacheDir: String?
        get() {
            val cacheDir = HabiticaBaseApplication.getInstance(context)?.getExternalFilesDir(Environment.DIRECTORY_NOTIFICATIONS)
            return cacheDir?.path
        }

    @SuppressLint("SetWorldReadable", "ReturnCount")
    suspend fun download(files: List<SoundFile>): List<SoundFile> {
        return files.map { audioFile ->
            withContext(Dispatchers.IO) {
                val file = File(getFullAudioFilePath(audioFile))
                if (file.exists() && file.length() > 5000) {
                    // Important, or else the MediaPlayer can't access this file
                    file.setReadable(true, false)
                    audioFile.file = file
                    return@withContext audioFile
                }
                val request = Request.Builder().url(audioFile.webUrl).build()

                val response: Response
                try {
                    response = client.newCall(request).execute()
                    if (!response.isSuccessful) {
                        throw IOException()
                    }
                } catch (io: IOException) {
                    return@withContext audioFile
                }

                try {
                    val sink = file.sink().buffer()
                    sink.writeAll(response.body!!.source())
                    sink.flush()
                    sink.close()
                } catch (io: IOException) {
                    return@withContext audioFile
                }

                file.setReadable(true, false)
                audioFile.file = file
                return@withContext audioFile
            }
        }
    }

    private fun getFullAudioFilePath(soundFile: SoundFile): String =
        externalCacheDir + File.separator + soundFile.filePath
}
